sdk/resource/os_unix_test.go

```diff
@@ -0,0 +1,135 @@
+	copy(buf.Release[:], "5.0.0")
+	copy(buf.Version[:], "#1 SMP Thu May 6 12:34:56 UTC 2021")
+	copy(buf.Machine[:], "x86_64")
+
+	return nil
+}
+
+func fakeUnameProviderWithError(buf *unix.Utsname) error {
+	return fmt.Errorf("Error invoking uname(2)")
+}
+
+func TestUname(t *testing.T) {
+	resource.SetUnameProvider(fakeUnameProvider)
+
+	uname, err := resource.Uname()
+
+	require.Equal(t, uname, "Mock OS DESKTOP-PC 5.0.0 #1 SMP Thu May 6 12:34:56 UTC 2021 x86_64")
+	require.NoError(t, err)
+
+	resource.SetDefaultUnameProvider()
+}
+
+func TestUnameError(t *testing.T) {
+	resource.SetUnameProvider(fakeUnameProviderWithError)
+
+	uname, err := resource.Uname()
+
+	require.Empty(t, uname)
+	require.Error(t, err)
+
+	resource.SetDefaultUnameProvider()
+}
+
+func TestCharsToString(t *testing.T) {
+	tt := []struct {
+		Name     string
+		Bytes    []byte
+		Expected string
+	}{
+		{"Empty array", []byte{}, ""},
+		{"Empty string (null terminated)", []byte{0x00}, ""},
+		{"Nonempty string (null terminated)", []byte{0x31, 0x32, 0x33, 0x00}, "123"},
+		{"Nonempty string (non-null terminated)", []byte{0x31, 0x32, 0x33}, "123"},
+		{"Nonempty string with values after null", []byte{0x31, 0x32, 0x33, 0x00, 0x34}, "123"},
+	}
+
+	for _, tc := range tt {
+		tc := tc
+
+		t.Run(tc.Name, func(t *testing.T) {
+			result := resource.CharsToString(tc.Bytes)
+			require.EqualValues(t, tc.Expected, result)
+		})
+	}
+}
+
+func TestGetFirstAvailableFile(t *testing.T) {
+	file1, _ := ioutil.TempFile("", "candidate_")
+	file2, _ := ioutil.TempFile("", "candidate_")
+
+	filename1, filename2 := file1.Name(), file2.Name()
+
+	tt := []struct {
+		Name             string
+		Candidates       []string
+		ExpectedFileName string
+		ExpectedErr      string
+	}{
+		{"Gets first, skip second candidate", []string{filename1, filename2}, filename1, ""},
+		{"Skips first, gets second candidate", []string{"does_not_exists", filename2}, filename2, ""},
+		{"Skips first, gets second, ignores third candidate", []string{"does_not_exists", filename2, filename1}, filename2, ""},
+		{"No candidates (empty slice)", []string{}, "", "no candidate file available: []"},
+		{"No candidates (nil slice)", nil, "", "no candidate file available: []"},
+		{"Single nonexisting candidate", []string{"does_not_exists"}, "", "no candidate file available: [does_not_exists]"},
+		{"Multiple nonexisting candidates", []string{"does_not_exists", "this_either"}, "", "no candidate file available: [does_not_exists this_either]"},
+	}
+
+	for _, tc := range tt {
+		tc := tc
+
+		t.Run(tc.Name, func(t *testing.T) {
+			file, err := resource.GetFirstAvailableFile(tc.Candidates)
+
+			filename := ""
+			if file != nil {
+				filename = file.Name()
+			}
+
+			errString := ""
+			if err != nil {
+				errString = err.Error()
+			}
+
+			require.Equal(t, tc.ExpectedFileName, filename)
+			require.Equal(t, tc.ExpectedErr, errString)
+		})
+	}
+
+	os.Remove(filename1)
+	os.Remove(filename2)
```

Comment:
I originally went with `os.CreateTemp` instead of `ioutil.TempFile`, but reverted back (725d90ff0a2a014ae84f6aed1cbe825f32b9fb64) when the CI failed for Go < 1.16.

Go 1.14 was still listed on the project's README when I started to work on this, but [it has been updated](https://github.com/open-telemetry/opentelemetry-go/commit/d20e7228302077549282f1c2002418c90f341204) since then to reflect only the last two versions of Go as supported, so I guess it's safe to assume we can use `t.TempDir()`.

I've updated the code based on your suggestion to use `t.TempDir()` here ea07b21725aaac9046993b5957d85d9a84248914, and realized I can use `t.Cleanup()` in a couple of other places (f94b783224117fed3b31ad0c385e1a19c7d7a5a5).

I will leave this thread open just in case a maintainer has something to comment on.